Content.Shared/Cabinet/ItemCabinetSystem.cs

```diff
@@ -0,0 +1,91 @@
+﻿using Content.Shared.Containers.ItemSlots;
+using Content.Shared.Interaction;
+using Content.Shared.Nutrition.Components;
+using Content.Shared.Nutrition.EntitySystems;
+using Robust.Shared.Containers;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Content.Shared.Cabinet;
+
+/// <summary>
+/// Controls ItemCabinet slot locking and visuals.
+/// </summary>
+public sealed class ItemCabinetSystem : EntitySystem
+{
+    [Dependency] private readonly ItemSlotsSystem _slots = default!;
+    [Dependency] private readonly OpenableSystem _openable = default!;
+    [Dependency] private readonly SharedAppearanceSystem _appearance = default!;
+
+    /// <inheritdoc/>
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<ItemCabinetComponent, ComponentStartup>(OnStartup);
+        SubscribeLocalEvent<ItemCabinetComponent, EntInsertedIntoContainerMessage>(OnContainerModified);
+        SubscribeLocalEvent<ItemCabinetComponent, EntRemovedFromContainerMessage>(OnContainerModified);
+        SubscribeLocalEvent<ItemCabinetComponent, OpenableOpenedEvent>(OnOpened);
+        SubscribeLocalEvent<ItemCabinetComponent, OpenableClosedEvent>(OnClosed);
+    }
+
+    private void OnStartup(Entity<ItemCabinetComponent> ent, ref ComponentStartup args)
+    {
+        // update at startup to avoid copy pasting locked: true and locked: false for each closed/open prototype
+        SetSlotLock(ent, !_openable.IsOpen(ent));
+
+        UpdateAppearance(ent);
+    }
```

Comment:
if you only ever want this to run once then you should use mapinit. Component startup / init can run multiple times which may cause some issues with this